# ncbi-datasets-pylib
### NCBI Datasets is a resource that lets you easily gather data from NCBI.
The Datasets API is still in alpha, and we're updating it often to add new functionality, iron out bugs and enhance usability.
For some larger downloads, you may want to download a [dehydrated bag](https://www.ncbi.nlm.nih.gov/datasets/docs/rehydrate/), and retrieve the individual data files at a later time.


This Python package is automatically generated by the [OpenAPI Generator](https://openapi-generator.tech) project:

- API version: v1
- Package version: 12.19.0
- Build package: org.openapitools.codegen.languages.PythonClientCodegen

## Requirements.

Python >= 3.6

## Installation & Usage
### Install latest package from pypi

To install the pre-built python package, create a virtualenv and use pip:

```sh
python -m venv ve/
source ve/bin/activate
pip install ncbi-datasets-pylib
```

## Getting Started

Please follow the [installation procedure](#installation--usage) and then run the following:

```python

import time
import ncbi.datasets.openapi
from pprint import pprint
from ncbi.datasets.openapi.api import gene_api
from ncbi.datasets.openapi.model.rpc_status import RpcStatus
from ncbi.datasets.openapi.model.v1_download_summary import V1DownloadSummary
from ncbi.datasets.openapi.model.v1_fasta import V1Fasta
from ncbi.datasets.openapi.model.v1_gene_dataset_request import V1GeneDatasetRequest
from ncbi.datasets.openapi.model.v1_gene_dataset_request_content_type import V1GeneDatasetRequestContentType
from ncbi.datasets.openapi.model.v1_gene_dataset_request_sort_field import V1GeneDatasetRequestSortField
from ncbi.datasets.openapi.model.v1_gene_match import V1GeneMatch
from ncbi.datasets.openapi.model.v1_gene_metadata import V1GeneMetadata
from ncbi.datasets.openapi.model.v1_organism import V1Organism
from ncbi.datasets.openapi.model.v1_organism_query_request_tax_rank_filter import V1OrganismQueryRequestTaxRankFilter
from ncbi.datasets.openapi.model.v1_ortholog_request_content_type import V1OrthologRequestContentType
from ncbi.datasets.openapi.model.v1_ortholog_set import V1OrthologSet
from ncbi.datasets.openapi.model.v1_sci_name_and_ids import V1SciNameAndIds
from ncbi.datasets.openapi.model.v1_sort_direction import V1SortDirection
# Defining the host is optional and defaults to https://api.ncbi.nlm.nih.gov/datasets/v1
# See configuration.py for a list of all supported configuration parameters.
configuration = ncbi.datasets.openapi.Configuration(
    host = "https://api.ncbi.nlm.nih.gov/datasets/v1"
)

# The client must configure the authentication and authorization parameters
# in accordance with the API server security policy.
# Examples for each auth method are provided below, use the example that
# satisfies your auth use case.

# Configure API key authorization: ApiKeyAuthHeader
configuration.api_key['ApiKeyAuthHeader'] = 'YOUR_API_KEY'

# Uncomment below to setup prefix (e.g. Bearer) for API key, if needed
# configuration.api_key_prefix['ApiKeyAuthHeader'] = 'Bearer'


# Enter a context with an instance of the API client
with ncbi.datasets.openapi.ApiClient(configuration) as api_client:
    # Create an instance of the API class
    api_instance = gene_api.GeneApi(api_client)
    gene_ids = [
        59067,
    ] # [int] | NCBI gene ids
include_annotation_type = [
        V1Fasta("FASTA_UNSPECIFIED"),
    ] # [V1Fasta] | Select additional types of annotation to include in the data package.  If unset, no annotation is provided. (optional)
fasta_filter = [
        "fasta_filter_example",
    ] # [str] | Limit the FASTA sequences in the datasets package to these transcript and protein accessions (optional)
filename = "ncbi_dataset.zip" # str | Output file name. (optional) (default to "ncbi_dataset.zip")

    try:
        # Get a gene dataset by gene ID
        api_response = api_instance.download_gene_package(gene_ids, include_annotation_type=include_annotation_type, fasta_filter=fasta_filter, filename=filename)
        pprint(api_response)
    except ncbi.datasets.openapi.ApiException as e:
        print("Exception when calling GeneApi->download_gene_package: %s\n" % e)
```

## Documentation for API Endpoints

All URIs are relative to *https://api.ncbi.nlm.nih.gov/datasets/v1*

Class | Method | HTTP request | Description
------------ | ------------- | ------------- | -------------
*GeneApi* | [**download_gene_package**](docs/GeneApi.md#download_gene_package) | **GET** /gene/id/{gene_ids}/download | Get a gene dataset by gene ID
*GeneApi* | [**download_gene_package_post**](docs/GeneApi.md#download_gene_package_post) | **POST** /gene/download | Get a gene dataset by POST
*GeneApi* | [**gene_download_summary_by_accession**](docs/GeneApi.md#gene_download_summary_by_accession) | **GET** /gene/accession/{accessions}/download_summary | Get gene download summary by RefSeq Accession
*GeneApi* | [**gene_download_summary_by_id**](docs/GeneApi.md#gene_download_summary_by_id) | **GET** /gene/id/{gene_ids}/download_summary | Get gene download summary by GeneID
*GeneApi* | [**gene_download_summary_by_post**](docs/GeneApi.md#gene_download_summary_by_post) | **POST** /gene/download_summary | Get gene download summary
*GeneApi* | [**gene_download_summary_by_tax_and_symbol**](docs/GeneApi.md#gene_download_summary_by_tax_and_symbol) | **GET** /gene/symbol/{symbols}/taxon/{taxon}/download_summary | Get gene download summary by gene symbol
*GeneApi* | [**gene_metadata_by_accession**](docs/GeneApi.md#gene_metadata_by_accession) | **GET** /gene/accession/{accessions} | Get gene metadata by RefSeq Accession
*GeneApi* | [**gene_metadata_by_id**](docs/GeneApi.md#gene_metadata_by_id) | **GET** /gene/id/{gene_ids} | Get gene metadata by GeneID
*GeneApi* | [**gene_metadata_by_post**](docs/GeneApi.md#gene_metadata_by_post) | **POST** /gene | Get gene metadata as JSON
*GeneApi* | [**gene_metadata_by_tax_and_symbol**](docs/GeneApi.md#gene_metadata_by_tax_and_symbol) | **GET** /gene/symbol/{symbols}/taxon/{taxon} | Get gene metadata by gene symbol
*GeneApi* | [**gene_metadata_stream_by_post**](docs/GeneApi.md#gene_metadata_stream_by_post) | **POST** /gene/stream | Get gene metadata
*GeneApi* | [**gene_orthologs_by_id**](docs/GeneApi.md#gene_orthologs_by_id) | **GET** /gene/id/{gene_id}/orthologs | Get gene orthologs by gene ID
*GeneApi* | [**gene_tax_name_query**](docs/GeneApi.md#gene_tax_name_query) | **GET** /gene/taxon_suggest/{taxon_query} | Get a list of taxonomy names and IDs found in the gene dataset given a partial taxonomic name
*GeneApi* | [**gene_tax_tree**](docs/GeneApi.md#gene_tax_tree) | **GET** /gene/taxon/{taxon}/tree | Get a taxonomic subtree by taxonomic identifier
*GenomeApi* | [**assembly_descriptors_by_accessions**](docs/GenomeApi.md#assembly_descriptors_by_accessions) | **GET** /genome/accession/{accessions} | Get genome metadata by accession
*GenomeApi* | [**assembly_descriptors_by_bioproject**](docs/GenomeApi.md#assembly_descriptors_by_bioproject) | **GET** /genome/bioproject/{accessions} | Get genome metadata by bioproject accession
*GenomeApi* | [**assembly_descriptors_by_taxon**](docs/GenomeApi.md#assembly_descriptors_by_taxon) | **GET** /genome/taxon/{taxon} | Get genome metadata by taxonomic identifier
*GenomeApi* | [**check_assembly_availability**](docs/GenomeApi.md#check_assembly_availability) | **GET** /genome/accession/{accessions}/check | Check the validity of genome accessions
*GenomeApi* | [**check_assembly_availability_post**](docs/GenomeApi.md#check_assembly_availability_post) | **POST** /genome/check | Check the validity of many genome accessions in a single request
*GenomeApi* | [**download_assembly_package**](docs/GenomeApi.md#download_assembly_package) | **GET** /genome/accession/{accessions}/download | Get a genome dataset by accession
*GenomeApi* | [**download_assembly_package_post**](docs/GenomeApi.md#download_assembly_package_post) | **POST** /genome/download | Get a genome dataset by post
*GenomeApi* | [**genome_download_summary**](docs/GenomeApi.md#genome_download_summary) | **GET** /genome/accession/{accessions}/download_summary | Preview genome dataset download
*GenomeApi* | [**genome_download_summary_by_post**](docs/GenomeApi.md#genome_download_summary_by_post) | **POST** /genome/download_summary | Preview genome dataset download by POST
*GenomeApi* | [**genome_metadata_by_post**](docs/GenomeApi.md#genome_metadata_by_post) | **POST** /genome | Get genome metadata by variety of identifiers
*GenomeApi* | [**genome_tax_name_query**](docs/GenomeApi.md#genome_tax_name_query) | **GET** /genome/taxon_suggest/{taxon_query} | Get a list of taxonomy names and IDs found in the assembly dataset given a partial taxonomic name
*GenomeApi* | [**genome_tax_tree**](docs/GenomeApi.md#genome_tax_tree) | **GET** /genome/taxon/{taxon}/tree | Get a taxonomic subtree by taxonomic identifier
*ProkaryoteApi* | [**download_prokaryote_gene_package**](docs/ProkaryoteApi.md#download_prokaryote_gene_package) | **GET** /protein/accession/{accessions}/download | Get a prokaryote gene dataset by RefSeq protein accession
*ProkaryoteApi* | [**download_prokaryote_gene_package_post**](docs/ProkaryoteApi.md#download_prokaryote_gene_package_post) | **POST** /protein/accession/download | Get a prokaryote gene dataset by RefSeq protein accession by POST
*TaxonomyApi* | [**tax_name_query**](docs/TaxonomyApi.md#tax_name_query) | **GET** /taxonomy/taxon_suggest/{taxon_query} | Get a list of taxonomy names and IDs given a partial taxonomic name
*TaxonomyApi* | [**taxonomy_filtered_subtree**](docs/TaxonomyApi.md#taxonomy_filtered_subtree) | **GET** /taxonomy/taxon/{taxons}/filtered_subtree | Use taxonomic identifiers to get a filtered taxonomic subtree
*TaxonomyApi* | [**taxonomy_filtered_subtree_post**](docs/TaxonomyApi.md#taxonomy_filtered_subtree_post) | **POST** /taxonomy/filtered_subtree | Use taxonomic identifiers to get a filtered taxonomic subtree by post
*TaxonomyApi* | [**taxonomy_metadata**](docs/TaxonomyApi.md#taxonomy_metadata) | **GET** /taxonomy/taxon/{taxons} | Use taxonomic identifiers to get taxonomic metadata
*TaxonomyApi* | [**taxonomy_metadata_post**](docs/TaxonomyApi.md#taxonomy_metadata_post) | **POST** /taxonomy | Use taxonomic identifiers to get taxonomic metadata by post
*VersionApi* | [**version**](docs/VersionApi.md#version) | **GET** /version | Retrieve service version
*VirusApi* | [**sars2_protein_download**](docs/VirusApi.md#sars2_protein_download) | **GET** /virus/taxon/sars2/protein/{proteins}/download | Download SARS-CoV-2 protein and CDS datasets by protein name
*VirusApi* | [**sars2_protein_summary**](docs/VirusApi.md#sars2_protein_summary) | **GET** /virus/taxon/sars2/protein/{proteins} | Summary of SARS-CoV-2 protein and CDS datasets by protein name
*VirusApi* | [**sars2_protein_table**](docs/VirusApi.md#sars2_protein_table) | **GET** /virus/taxon/sars2/protein/{proteins}/table | Get SARS-CoV-2 protein metadata in a tabular format.
*VirusApi* | [**virus_genome_download**](docs/VirusApi.md#virus_genome_download) | **GET** /virus/taxon/{taxon}/genome/download | Download Coronavirus genome datasets by taxon
*VirusApi* | [**virus_genome_summary**](docs/VirusApi.md#virus_genome_summary) | **GET** /virus/taxon/{taxon}/genome | Get summary data for Coronaviridae genomes by taxon
*VirusApi* | [**virus_genome_table**](docs/VirusApi.md#virus_genome_table) | **GET** /virus/taxon/{taxon}/genome/table | Get viral genomic metadata in a tabular format.


## Documentation For Models

 - [ProtobufAny](docs/ProtobufAny.md)
 - [RpcStatus](docs/RpcStatus.md)
 - [V1Accessions](docs/V1Accessions.md)
 - [V1AnnotatedAssemblies](docs/V1AnnotatedAssemblies.md)
 - [V1Annotation](docs/V1Annotation.md)
 - [V1AnnotationForAssembly](docs/V1AnnotationForAssembly.md)
 - [V1AnnotationForAssemblyFile](docs/V1AnnotationForAssemblyFile.md)
 - [V1AnnotationForAssemblyType](docs/V1AnnotationForAssemblyType.md)
 - [V1AnnotationForVirusType](docs/V1AnnotationForVirusType.md)
 - [V1AssemblyDatasetAvailability](docs/V1AssemblyDatasetAvailability.md)
 - [V1AssemblyDatasetDescriptor](docs/V1AssemblyDatasetDescriptor.md)
 - [V1AssemblyDatasetDescriptorChromosome](docs/V1AssemblyDatasetDescriptorChromosome.md)
 - [V1AssemblyDatasetDescriptors](docs/V1AssemblyDatasetDescriptors.md)
 - [V1AssemblyDatasetDescriptorsFilter](docs/V1AssemblyDatasetDescriptorsFilter.md)
 - [V1AssemblyDatasetDescriptorsFilterAssemblyLevel](docs/V1AssemblyDatasetDescriptorsFilterAssemblyLevel.md)
 - [V1AssemblyDatasetDescriptorsFilterAssemblySource](docs/V1AssemblyDatasetDescriptorsFilterAssemblySource.md)
 - [V1AssemblyDatasetDescriptorsRequest](docs/V1AssemblyDatasetDescriptorsRequest.md)
 - [V1AssemblyDatasetDescriptorsRequestContentType](docs/V1AssemblyDatasetDescriptorsRequestContentType.md)
 - [V1AssemblyDatasetRequest](docs/V1AssemblyDatasetRequest.md)
 - [V1AssemblyDatasetRequestResolution](docs/V1AssemblyDatasetRequestResolution.md)
 - [V1AssemblyMatch](docs/V1AssemblyMatch.md)
 - [V1AssemblyMetadata](docs/V1AssemblyMetadata.md)
 - [V1AssemblyMetadataRequest](docs/V1AssemblyMetadataRequest.md)
 - [V1AssemblyMetadataRequestBioprojects](docs/V1AssemblyMetadataRequestBioprojects.md)
 - [V1AssemblyMetadataRequestContentType](docs/V1AssemblyMetadataRequestContentType.md)
 - [V1BioProject](docs/V1BioProject.md)
 - [V1BioProjectLineage](docs/V1BioProjectLineage.md)
 - [V1BuscoStat](docs/V1BuscoStat.md)
 - [V1CountType](docs/V1CountType.md)
 - [V1DatasetRequest](docs/V1DatasetRequest.md)
 - [V1DownloadSummary](docs/V1DownloadSummary.md)
 - [V1DownloadSummaryAvailableFiles](docs/V1DownloadSummaryAvailableFiles.md)
 - [V1DownloadSummaryDehydrated](docs/V1DownloadSummaryDehydrated.md)
 - [V1DownloadSummaryFileSummary](docs/V1DownloadSummaryFileSummary.md)
 - [V1DownloadSummaryHydrated](docs/V1DownloadSummaryHydrated.md)
 - [V1Error](docs/V1Error.md)
 - [V1ErrorAssemblyErrorCode](docs/V1ErrorAssemblyErrorCode.md)
 - [V1ErrorGeneErrorCode](docs/V1ErrorGeneErrorCode.md)
 - [V1ErrorVirusErrorCode](docs/V1ErrorVirusErrorCode.md)
 - [V1Fasta](docs/V1Fasta.md)
 - [V1FeatureCounts](docs/V1FeatureCounts.md)
 - [V1GeneCounts](docs/V1GeneCounts.md)
 - [V1GeneDatasetRequest](docs/V1GeneDatasetRequest.md)
 - [V1GeneDatasetRequestContentType](docs/V1GeneDatasetRequestContentType.md)
 - [V1GeneDatasetRequestSort](docs/V1GeneDatasetRequestSort.md)
 - [V1GeneDatasetRequestSortField](docs/V1GeneDatasetRequestSortField.md)
 - [V1GeneDatasetRequestSymbolsForTaxon](docs/V1GeneDatasetRequestSymbolsForTaxon.md)
 - [V1GeneDescriptor](docs/V1GeneDescriptor.md)
 - [V1GeneDescriptorGeneType](docs/V1GeneDescriptorGeneType.md)
 - [V1GeneDescriptorRnaType](docs/V1GeneDescriptorRnaType.md)
 - [V1GeneDescriptors](docs/V1GeneDescriptors.md)
 - [V1GeneMatch](docs/V1GeneMatch.md)
 - [V1GeneMetadata](docs/V1GeneMetadata.md)
 - [V1GenomicLocation](docs/V1GenomicLocation.md)
 - [V1GenomicRegion](docs/V1GenomicRegion.md)
 - [V1GenomicRegionGenomicRegionType](docs/V1GenomicRegionGenomicRegionType.md)
 - [V1MaturePeptide](docs/V1MaturePeptide.md)
 - [V1Message](docs/V1Message.md)
 - [V1MethodPayloadRequest](docs/V1MethodPayloadRequest.md)
 - [V1MicroBiggeDatasetRequest](docs/V1MicroBiggeDatasetRequest.md)
 - [V1MicroBiggeDatasetRequestElementFlankConfig](docs/V1MicroBiggeDatasetRequestElementFlankConfig.md)
 - [V1MicroBiggeDatasetRequestFileType](docs/V1MicroBiggeDatasetRequestFileType.md)
 - [V1NomenclatureAuthority](docs/V1NomenclatureAuthority.md)
 - [V1Organism](docs/V1Organism.md)
 - [V1OrganismCountByType](docs/V1OrganismCountByType.md)
 - [V1OrganismCounts](docs/V1OrganismCounts.md)
 - [V1OrganismQueryRequest](docs/V1OrganismQueryRequest.md)
 - [V1OrganismQueryRequestTaxRankFilter](docs/V1OrganismQueryRequestTaxRankFilter.md)
 - [V1OrganismRankType](docs/V1OrganismRankType.md)
 - [V1Orientation](docs/V1Orientation.md)
 - [V1OrthologRequest](docs/V1OrthologRequest.md)
 - [V1OrthologRequestContentType](docs/V1OrthologRequestContentType.md)
 - [V1OrthologSet](docs/V1OrthologSet.md)
 - [V1ProkaryoteGeneRequest](docs/V1ProkaryoteGeneRequest.md)
 - [V1ProkaryoteGeneRequestGeneFlankConfig](docs/V1ProkaryoteGeneRequestGeneFlankConfig.md)
 - [V1Protein](docs/V1Protein.md)
 - [V1Range](docs/V1Range.md)
 - [V1Sars2ProteinDatasetRequest](docs/V1Sars2ProteinDatasetRequest.md)
 - [V1SciNameAndIds](docs/V1SciNameAndIds.md)
 - [V1SciNameAndIdsSciNameAndId](docs/V1SciNameAndIdsSciNameAndId.md)
 - [V1SeqRangeSet](docs/V1SeqRangeSet.md)
 - [V1SleepReply](docs/V1SleepReply.md)
 - [V1SleepRequest](docs/V1SleepRequest.md)
 - [V1SortDirection](docs/V1SortDirection.md)
 - [V1TableFormat](docs/V1TableFormat.md)
 - [V1TabularOutput](docs/V1TabularOutput.md)
 - [V1TaxTreeRequest](docs/V1TaxTreeRequest.md)
 - [V1TaxidReply](docs/V1TaxidReply.md)
 - [V1TaxidRequest](docs/V1TaxidRequest.md)
 - [V1TaxonomyFilteredSubtreeRequest](docs/V1TaxonomyFilteredSubtreeRequest.md)
 - [V1TaxonomyFilteredSubtreeResponse](docs/V1TaxonomyFilteredSubtreeResponse.md)
 - [V1TaxonomyFilteredSubtreeResponseEdge](docs/V1TaxonomyFilteredSubtreeResponseEdge.md)
 - [V1TaxonomyFilteredSubtreeResponseEdgeChildStatus](docs/V1TaxonomyFilteredSubtreeResponseEdgeChildStatus.md)
 - [V1TaxonomyFilteredSubtreeResponseEdgesEntry](docs/V1TaxonomyFilteredSubtreeResponseEdgesEntry.md)
 - [V1TaxonomyMatch](docs/V1TaxonomyMatch.md)
 - [V1TaxonomyMetadataRequest](docs/V1TaxonomyMetadataRequest.md)
 - [V1TaxonomyMetadataRequestContentType](docs/V1TaxonomyMetadataRequestContentType.md)
 - [V1TaxonomyMetadataResponse](docs/V1TaxonomyMetadataResponse.md)
 - [V1TaxonomyNode](docs/V1TaxonomyNode.md)
 - [V1TaxonomyNodeCountByType](docs/V1TaxonomyNodeCountByType.md)
 - [V1Transcript](docs/V1Transcript.md)
 - [V1TranscriptTranscriptType](docs/V1TranscriptTranscriptType.md)
 - [V1VersionReply](docs/V1VersionReply.md)
 - [V1VirusDatasetRequest](docs/V1VirusDatasetRequest.md)
 - [V1VirusDatasetRequestAccessions](docs/V1VirusDatasetRequestAccessions.md)
 - [V1VirusTableField](docs/V1VirusTableField.md)
 - [V1Warning](docs/V1Warning.md)
 - [V1WarningGeneWarningCode](docs/V1WarningGeneWarningCode.md)
 - [V1WarningReplacedId](docs/V1WarningReplacedId.md)
 - [V1reportsGeneDescriptorGeneType](docs/V1reportsGeneDescriptorGeneType.md)
 - [V1reportsGeneDescriptorRnaType](docs/V1reportsGeneDescriptorRnaType.md)
 - [V1reportsGenomicRegionGenomicRegionType](docs/V1reportsGenomicRegionGenomicRegionType.md)
 - [V1reportsOrientation](docs/V1reportsOrientation.md)
 - [V1reportsProkaryoteGeneLocationCompleteness](docs/V1reportsProkaryoteGeneLocationCompleteness.md)
 - [V1reportsPurposeOfSampling](docs/V1reportsPurposeOfSampling.md)
 - [V1reportsTranscriptTranscriptType](docs/V1reportsTranscriptTranscriptType.md)
 - [V1reportsVirusAssemblyCompleteness](docs/V1reportsVirusAssemblyCompleteness.md)


## Documentation For Authorization


## ApiKeyAuth

- **Type**: API key
- **API key parameter name**: api_key
- **Location**: URL query string


## ApiKeyAuthHeader

- **Type**: API key
- **API key parameter name**: api-key
- **Location**: HTTP header


## Author




## Notes for Large OpenAPI documents
If the OpenAPI document is large, imports in ncbi.datasets.openapi.apis and ncbi.datasets.openapi.models may fail with a
RecursionError indicating the maximum recursion limit has been exceeded. In that case, there are a couple of solutions:

Solution 1:
Use specific imports for apis and models like:
- `from ncbi.datasets.openapi.api.default_api import DefaultApi`
- `from ncbi.datasets.openapi.model.pet import Pet`

Solution 2:
Before importing the package, adjust the maximum recursion limit as shown below:
```
import sys
sys.setrecursionlimit(1500)
import ncbi.datasets.openapi
from ncbi.datasets.openapi.apis import *
from ncbi.datasets.openapi.models import *
```

